[WIP] ICMP: Option to disable checking checksum

cfriedt opened the pull request:
Previously, the samples/bluetooth/ipsp example did not work.

The following appeared after enabling network debug logging.

[00:05:42.240,753] <dbg> net_ipv6.net_ipv6_input: (0x20001e48): IPv6 packet len 72 received from fe80::5ef3:70ff:fe92:5c06 to ff02::1:ffe9:6fd0
[00:05:42.240,814] <dbg> net_icmpv6.net_icmpv6_input: (0x20001e48): DROP: invalid checksum

As @tbursztyka theorized in Slack, Linux was producing a correct ICMPv6 checksum, as shown by Wireshark (this was taken after my change was applied).
![wireshark-zephyr-icmpv6-checksum](https://user-images.githubusercontent.com/983494/66492868-18d8fd80-ea83-11e9-9f41-2965bbd77ac7.png)

After inspecting the code, indeed it seems to be the case that if net_calc_chksum_icmpv6(pkt) != 0U, then the packet is dropped.

It's possible that 6LowPan code (either in Linux or Zephyr) is not handling the IPv6 checksum properly, but I have not investigated any further.

Ideally, there would be an easy way to enable ICMPv6 checksum in Linux via /proc/sys/net/ipv6/icmp but the only node that exists there (at least on my desk) is ratelimit. If there were an easy way to enable it, then the Zephyr documentation could simply be modified.

It is a bit of a blocker to simply run a demo application. However, as @jukkar pointed out in Slack, it would be desirable to keep the checksum by default but allow it to be disabled (independently) for ICMPv6 and ICMPv4.

Therefore, I'm proposing to add a Kconfig variable to disable ICMP checksum verification and to explicitly set it to n for samples/bluetooth/ipsp.

That would also ensure that both the example and documentation remains consistent as well as the current behaviour of the ICMPv4 and ICMPv6 code.

jukkar commented:
The commit message is quite verbose, we need to probably tweak it a bit.

pfalcon commented:
> Previously, the samples/bluetooth/ipsp example did not work.

I don't remember the issues described when I tested it (quite some time ago though). So for me, this patch suspiciously looks like "fixing" misconfiguration of a particular Linux system by patching away Zephyr functionality.

cfriedt commented:
> > Previously, the samples/bluetooth/ipsp example did not work.
> 
> I don't remember the issues described when I tested it (quite some time ago though). So for me, this patch suspiciously looks like "fixing" misconfiguration of a particular Linux system by patching away Zephyr functionality.

It's actually several million particular Linux systems - I'm just using Ubuntu 18.04 (LTS). However, yeah, I believe it probably worked fine in the past and could be a regression. Also, some network stacks (e.g. NetX) genuinely do not do checksumming OOTB on embedded platforms, so it could be a useful "feature" in that sense.

pfalcon commented:
> It's actually several million particular Linux systems - I'm just using Ubuntu 18.04 (LTS).

While @jukkar suggest to prune commit message, I'd suggest  vice-versa, to include that information to the commit message, so whoever will deal with that matter in the future, had a good understanding why it was done. For reference, I did my testing yet with ubuntu 16.04, and default kernel didn't really work for 6lowpan/BLE, so I had to upgrade it to some newer version.

> so it could be a useful "feature" in that sense.

Sounds good. My concern is that you propose to enable it (disable checksum calc) by default in the reference IPSP tool, yet admit yourself that you didn't do deep enough investigation of it. (At ubuntu 18.04 ref sheds some light at the situation.)

jukkar commented:
For the commit message I propose something like this:
```
The samples/bluetooth/ipsp example did not work as expected and
following output appeared after enabling network debug logging.

[00:05:42.240,753] <dbg> net_ipv6.net_ipv6_input: (0x20001e48): IPv6 packet len 72 received from fe80::5ef3:70ff:fe92:5c06 to ff02::1:ffe9:6fd0
[00:05:42.240,814] <dbg> net_icmpv6.net_icmpv6_input: (0x20001e48): DROP: invalid checksum

In order to accept the ICMP packet regardless of the checksum value,
add new Kconfig options that can be used to disable ICMP checksum
checks.
```

cfriedt commented:
> > It's actually several million particular Linux systems - I'm just using Ubuntu 18.04 (LTS).
> 
> While @jukkar suggest to prune commit message, I'd suggest vice-versa, to include that information to the commit message, so whoever will deal with that matter in the future, had a good understanding why it was done. For reference, I did my testing yet with ubuntu 16.04, and default kernel didn't really work for 6lowpan/BLE, so I had to upgrade it to some newer version.

I'm ok with a longer commit message as well, just to shed some light. Are you ok with that @jukkar ?

> > so it could be a useful "feature" in that sense.
> 
> Sounds good. My concern is that you propose to enable it (disable checksum calc) by default in the reference IPSP tool, yet admit yourself that you didn't do deep enough investigation of it. (At ubuntu 18.04 ref sheds some light at the situation.)

Well.. like I said in Slack, the likelihood that (Linux OR iputils) AND Wireshark both have simultaneously messed up their ICMPv6 checksum calculation is very slim. 

I hate pointing fingers, but think that it might be a regression in Zephyr.

There is a good way to check - unit testing :smile: 

It's also possible that a unit test for the ICMPv6 checksum could pass, but that an integration test for the RX chain could fail.

But I'm not familiar enough with Zephyr's test infra to say whether it is capable of integration tests for the entire RX chain going through multiple subsystems (BLE or 15.4 / 6LowPan / IP / ICMPv6).

jfischer-phytec-iot commented:
Do I understand it correctly, there is a bug somewhere (in 6lo decompression?) and instead to fix it ICMPv6 checksum check for the IPSP sample will be turned off ?

cfriedt commented:
> Do I understand it correctly, there is a bug somewhere (in 6lo decompression?) and instead to fix it ICMPv6 checksum check for the IPSP sample will be turned off ?

The current _theory_ is that yes.

I have not found the root cause though, and only suggest adding the Kconfig option to disable checksum verification (the default would still be to enable checksum verification).

jfischer-phytec-iot commented:
> > Do I understand it correctly, there is a bug somewhere (in 6lo decompression?) and instead to fix it ICMPv6 checksum check for the IPSP sample will be turned off ?
> 
> The current _theory_ is that yes.
> 
> I have not found the root cause though, and only suggest adding the Kconfig option to disable checksum verification (the default would still be to enable checksum verification).

Default is disabled for IPSP 902005dc7c3112381dc35b737e17f21baf134ec1. It is wrong. The bug should be fixed.

jfischer-phytec-iot commented on samples/bluetooth/ipsp/prj.conf:
```diff
@@ -10,6 +10,7 @@ CONFIG_NET_IPV6=y
 CONFIG_NET_IPV4=n
 CONFIG_NET_UDP=y
 CONFIG_NET_TCP=y
+CONFIG_NET_ICMPV6_CHECKSUM=n
```
This is not right at all. The bug should be fixed.

cfriedt commented:
OK - so it looks like the issue is now gone. It was possible that it was weird temporarily unstable HEAD. I did rebase at one point and saw some changes go in by @cvinayak afterward. 

Feel free to close, or if you would like to keep the change that allows checksums to be disabled, I can remove the change for the ipsp example.

tbursztyka commented:
so bluetooth was somehow broken then? (which is a valid possibility actually)

cfriedt commented:
> so bluetooth was somehow broken then? (which is a valid possibility actually)

¯\_(ツ)_/¯ - maybe. It works now though. Sorry for the interruption.

cfriedt committed:
net: icmp: option to disable checking checksum

Briefly, the samples/bluetooth/ipsp example did not work. It seems to be working now, but I found the following appeared after enabling network debug logging.

[00:05:42.240,753] net_ipv6.net_ipv6_input: (0x20001e48): IPv6 packet len 72 received from fe80::5ef3:70ff:fe92:5c06 to ff02::1:ffe9:6fd0
[00:05:42.240,814] net_icmpv6.net_icmpv6_input: (0x20001e48): DROP: invalid checksum

As @tbursztyka theorized in Slack, Linux was producing a correct ICMPv6 checksum, as shown by Wireshark (this was taken after my change was applied).

After inspecting the code, indeed it seems to be the case that if net_calc_chksum_icmpv6(pkt) != 0U, then the packet is dropped.

It's possible that 6LowPan code (either in Linux or Zephyr) is not handling the IPv6 checksum properly, but I have not investigated any further.

Ideally, there would be an easy way to enable ICMPv6 checksum in Linux via /proc/sys/net/ipv6/icmp but the only node that exists there (at least on my desk) is ratelimit. If there were an easy way to enable it, then the Zephyr documentation could simply be modified.

It is a bit of a blocker to simply run a demo application. However, as @jukkar pointed out in Slack, it would be desirable to keep the checksum by default but allow it to be disabled (independently) for ICMPv6 and ICMPv4.

Therefore, I'm proposing to add a Kconfig variable to disable ICMP checksum verification and to explicitly set it to n for samples/bluetooth/ipsp.

That would also ensure that both the example and documentation remains consistent as well as the current behaviour of the ICMPv4 and ICMPv6 code.

cfriedt commented:
My first priority is getting a new BLE platform in, so when that's done, I would be happy to help out in any way I can :-)

I'll close this PR / issue for now, if that's ok with everyone.
